### PR TITLE
cudatoolkit_11_{3,4}: init at 11.{3,4}.1

### PR DESCRIPTION
**Pull request by samuela**
###### Motivation for this change
Add derivations for the latest CUDA toolkit releases.

GCC versions were selected based on the official documentation 
* 11.3: https://docs.nvidia.com/cuda/archive/11.3.1/cuda-installation-guide-linux/index.html#system-requirements
* 11.4: https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#system-requirements

I have built both derivations and tested that they run `nvidia-smi` successfully.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (r-burns) [inline pkgs/development/compilers/cudatoolkit/default.nix:74]:
```suggestion
  cudatoolkit_11 = cudatoolkit_11_4;
```

**Review by r-burns — COMMENTED** [review]
I think there are a couple other places that need to be updated such as pkgs/test/cuda

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (danieldk) [reply pkgs/development/compilers/cudatoolkit/default.nix:71]: If I recall correctly using newer versions also sometimes gives errors, most distributions that they are targeting have older gcc versions. I think the best way of finding out is to build a reasonably large subset our packages that support CUDA 11 against this version and see if there are any issues.

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (danieldk) [reply pkgs/development/compilers/cudatoolkit/default.nix:74]: I think we should add a [compatible CuDNN](https://docs.nvidia.com/deeplearning/cudnn/support-matrix/index.html) first before updating `cudatoolkit_11`, since otherwise `cudatoolkit_11` and  `cudnn_cudatoolkit_11` are not compatible.

**Review by danieldk — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (samuela) [reply pkgs/development/compilers/cudatoolkit/default.nix:74]: I'd be happy to do that as well. Either in this pr or in a follow-up one.

**Comment by samuela** [comment]
@domenkozar I noticed that you reviewed as "request changes" regarding the `cudatoolkit_11` alias. I actually wasn't planning on updating the `cudatoolkit_11` alias in this PR since it may have cudnn/etc conflicts. What are people's preferences on updating everything in this PR vs updating in future PRs? I'm happy to do either one.

> I think there are a couple other places that need to be updated such as pkgs/test/cuda

Ah, good call. This may be a noob question but how do I run the tests in pkgs/test/cuda?

**Comment by r-burns** [comment]
Not sure about updating everything at once, it looks like the cudatoolkits alone have been bumped in the past. I'll defer to @danieldk on this

The package tests are exposed in the `tests` attrset, e.g. `tests.cuda.cuda-library-samples_cudatoolkit_11_4`

**Comment by danieldk** [comment]
> I actually wasn't planning on updating the `cudatoolkit_11` alias in this PR since it may have cudnn/etc conflicts.

I'd say, don't update it in this PR, just introduce the new toolkit versions. Then a separate PR for cuDNN and then finally a third (or added to the cuDNN PR) to update the aliases.

IMO it also only makes sense to update the aliases after checking that major libraries (PyTorch, Tensorflow if it buidls) work correctly. It would be bad if something major like PyTorch doesn't work with our default CUDA 11.

**Comment by samuela** [comment]
Ok I added tests for the new cudatoolkit versions in https://github.com/NixOS/nixpkgs/pull/136655/commits/803b04246498f2731ead99d22703fdcf5a16f31b. Some of the tests relied on cutensor which required adding new cutensor versions in https://github.com/NixOS/nixpkgs/pull/136655/commits/6f9f432136bd5e1515479348cc416d77a887fcab.

The builds of both of all new tests -- cuda-samples_cudatoolkit_11_3, cuda-samples_cudatoolkit_10_4, cuda-library-samples_cudatoolkit_11_3, and cuda-library-samples_cudatoolkit_11_4 -- are passing. However, I noticed that the `cuda-samples_cudatoolkit_10_2` has a pre-existing failure:
```
$ nix-build -A tests.cuda.cuda-samples_cudatoolkit_10_2
...
strip is /nix/store/v8imx1nvyz0hgvx9cbcmh6gp4ngw3ffj-binutils-2.35.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/x0n95ddxf00v1ffzjl88k00nm00z1nvz-cuda-samples-10.2/bin
patching script interpreter paths in /nix/store/x0n95ddxf00v1ffzjl88k00nm00z1nvz-cuda-samples-10.2
checking for references to /build/ in /nix/store/x0n95ddxf00v1ffzjl88k00nm00z1nvz-cuda-samples-10.2...
patchelf: unsupported overlap of SHT_NOTE and PT_NOTE
builder for '/nix/store/fafbfzyqw0v8in3cfl8c1ydr7bazvk8y-cuda-samples-10.2.drv' failed with exit code 1
error: build of '/nix/store/fafbfzyqw0v8in3cfl8c1ydr7bazvk8y-cuda-samples-10.2.drv' failed
```

I have not updated the `cutensor_cudatoolkit_11` alias since python3Packages.cupy, the only reverse dependency of cutensor_cudatoolkit_11, is currently at v8.4.0 which supports at most cutensor v1.2. Upgrading cupy to the latest, v9.4, should resolve that issue.

As a side note, how are tests that require CUDA/GPU access handled in the nix testing ecosystem? I've found that the nix build environment blocks access to the GPU, which really inhibits writing tests for GPU-enabled code.

I think this PR should be more or less ready to go now.

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/cudatoolkit/default.nix:63]: Please remove the sha256: prefix.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (samuela) [reply pkgs/development/compilers/cudatoolkit/default.nix]: Done in https://github.com/NixOS/nixpkgs/pull/136655/commits/09b33fe7b4412dad09a7aecd95bbe54364b71567.

**Comment by r-burns** [comment]
Did we agree to use gcc10 for cudatoolkit 11.4? Besides that LGTM. All the samples I tried out work on my machine.

**Comment by r-burns** [comment]
> As a side note, how are tests that require CUDA/GPU access handled in the nix testing ecosystem? I've found that the nix build environment blocks access to the GPU, which really inhibits writing tests for GPU-enabled code.

Yeah I think currently there is no testing on GPU, it all has to happen out-of-band on the resulting packages which isn't ideal. Locally I've found that I can add /run/opengl-driver/lib, /dev/nvidia0, /dev/nvidiactl, etc to extra-sandbox-paths to enable this, and I think it would be good to standardize this in nixpkgs if possible. I think nvcc automatically builds native code for detected GPUs though, so for reproducibility we might need to watch out for that.

**Comment by samuela** [comment]
> Did we agree to use gcc10 for cudatoolkit 11.4? Besides that LGTM. All the samples I tried out work on my machine.

I picked gcc versions according to the docs [here](https://docs.nvidia.com/cuda/archive/11.3.1/cuda-installation-guide-linux/index.html#system-requirements) and [here](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#system-requirements). So according to NVIDIA they _should_ be happy combos. If we come across issues in updating the `cudatoolkit_11` alias, it should be pretty easy to detect any issues.

**Comment by r-burns** [comment]
Right, I'm just worried about the potential linking issues when mixing libstdc++ versions between the stdenv.cc and cuda-host-compiler. For older toolkits our hands are tied, but cudatoolkit.cc = stdenv.cc should be the most reliable combo when permissible.

(Hopefully someday once the gcc packaging is fully modularized, we can rewrap cudatoolkit.cc and stdenv.cc to use the same libstdc++)

**Comment by samuela** [comment]
I don't know enough about the libstdc++ and stdenv situation in NixOS to have an intelligent response. But I'm happy to make whatever changes that are collectively decided to make the most sense. I picked the latest compatible gcc's with the rationale that newer is better.

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (r-burns) [inline pkgs/development/compilers/cudatoolkit/default.nix:71]:
```suggestion
    gcc = gcc10; # can bump to 11 along with stdenv.cc
```
Ok let's stick with this for now. LGTM otherwise.

**Review by r-burns — CHANGES_REQUESTED** [review]

**Review by r-burns — APPROVED** [review]
Looks good, thanks!

**Comment by samuela** [comment]
I just rebased against master (currently 69a9d335f633ef221327821089877c98004b70bc) in order to get the new cupy version (upgraded in https://github.com/NixOS/nixpkgs/commit/9c4a68a2519842d78543ef5cdb2ffb236d8a80e1). Unfortunately however I seem to have discovered a patchelf regression in the process. Without changing anything in the PR diff, on the latest master, I'm seeing a patchelf segfault that is reproducible:
```
...
patching /nix/store/030p8ynsjicyyiifajxkyzhmyrni7gnl-cudatoolkit-11.3.1/bin/ptxas...
find: '/nix/store/432xc7hrzjh6p49lh49axgk87zggpbsg-cudatoolkit-11.3.1-doc': No such file or directory
patching /nix/store/lvqc5baxznz5a5plhnjkv23rywl9q8nh-cudatoolkit-11.3.1-lib/lib/libcudart.so.11.3.109...
/nix/store/9jsp95c5r4pl1pdqzy9z405wj39czpwa-stdenv-linux/setup: line 100:  2286 Segmentation fault      (core dumped) patchelf --set-rpath "$rpath2" --force-rpath $i
builder for '/nix/store/6c5h63rd5lakmc43wfzc147806rysklp-cudatoolkit-11.3.1.drv' failed with exit code 139
```
Has anyone seen anything like this before?

**Comment by r-burns** [comment]
Try this fix maybe? https://github.com/NixOS/nixpkgs/pull/138231

**Comment by samuela** [comment]
@r-burns Yes! That does the trick. So I guess let's get https://github.com/NixOS/nixpkgs/pull/138231 merged, and then I'll rebase this once that's merged?

File: pkgs/development/compilers/cudatoolkit/default.nix
Comment (r-burns) [inline pkgs/development/compilers/cudatoolkit/default.nix:6]:
```suggestion
, gcc10
```
Oops, missed this

Comment (samuela) [reply pkgs/development/compilers/cudatoolkit/default.nix]: So did I! Should be fixed in https://github.com/NixOS/nixpkgs/commit/0981d9473f646678a7279df30530c5ba51519ebe

**Comment by samuela** [comment]
> @r-burns Yes! That does the trick. So I guess let's get #138231 merged, and then I'll rebase this once that's merged?

All done, I rebased in https://github.com/NixOS/nixpkgs/commit/0981d9473f646678a7279df30530c5ba51519ebe.

**Review by r-burns — APPROVED** [review]
Nice, toolkits build on the rebased tip.

File: pkgs/development/libraries/science/math/cutensor/default.nix
Comment (r-burns) [inline pkgs/development/libraries/science/math/cutensor/default.nix:44]:
```suggestion
  cutensor_cudatoolkit_11 = cutensor_cudatoolkit_11_2;
```
I just discovered that this would bloat the closure of referrers which also use cudnn_cudatoolkit_11, such as cupy.
Let's plan on a big flag day bump for cudnn/cutensor/cudatoolkit -> 11.4.
Exception being the samples, which I think are fine to bump here since nothing depends on them.

Comment (samuela) [reply pkgs/development/libraries/science/math/cutensor/default.nix]: Ah good catch, applied your suggestion!

**Comment by r-burns** [comment]
@danieldk requesting a final skim if you have a moment. Have tested thoroughly and should be good to go AFAICT but would appreciate an ACK from someone more experienced.

**Comment by r-burns** [comment]
I'm gonna merge this in a couple hours if no one objects. I'll be available for damage control for the rest of the day in case something goes wrong.

**Comment by r-burns** [comment]
@samuela thanks for your hard work and patience :heart:

**Comment by samuela** [comment]
@r-burns no problem, happy to help! And thanks for all your assistance getting this reviewed and merged!

**Pull request merged**
